### PR TITLE
[Ledger] Sign BNB transactions

### LINKED ISSUE #1772: [Ledger] Add business logic to send / sign BNB transactions
Check previous PR #670 about same thing. Have also a look into similar PR for THORChain #1714 (which is different comparing to BNB)

Part of #1709

### PR DESCRIPTION
**Pull request by gromxyz**
Closes #1772 #1784

### CODE REVIEW

File: src/renderer/services/binance/transaction.ts [hunk 1]
```diff
@@ -1,4 +1,66 @@
...
+export const createTransactionService = (client$: Client$, network$: Network$): TransactionService => {
+  const common = C.createTransactionService(client$)
+
+  const sendLedgerTx = ({ network, params }: { network: Network; params: SendTxParams }) => {
+    const sendLedgerTxParams: IPCLedgerSendTxParams = {
+      chain: BNBChain,
+      network,
+      asset: params.asset,
+      amount: params.amount,
+      sender: params.sender ? params.sender : '',
```
Comment (veado) [inline src/renderer/services/binance/transaction.ts]:
```ts
sender: params.sender
```

File: src/main/api/ledger/binance/transaction.ts [hunk 2]
```diff
@@ -1,53 +1,89 @@
...
+        errorId: LedgerErrorId.INVALID_RESPONSE,
+        msg: `Binance client failed to send transaction`
+      })
+    }
+    const txhash = result[0]['hash']
+
+    if (!txhash) {
+      return E.left({
+        errorId: LedgerErrorId.INVALID_RESPONSE,
+        msg: `Post request to send 'MsgSend' failed`
```
Comment (veado) [inline src/main/api/ledger/binance/transaction.ts]:
```ts
`Post request to send ${asset?.symbol ?? AssetBNB.symbol} transaction by using Ledger failed`
```

File: src/main/api/ledger/binance/transaction.ts [hunk 3]
```diff
@@ -1,53 +1,89 @@
...
+      recipient,
+      baseToAsset(amount).amount().toString(),
+      asset ? asset.symbol : AssetBNB.symbol,
+      memo
+    )
+
+    if (result.length === 0 || result[0].hash === undefined) {
+      return E.left({
+        errorId: LedgerErrorId.INVALID_RESPONSE,
+        msg: `Binance client failed to send transaction`
```
Comment (veado) [inline src/main/api/ledger/binance/transaction.ts]:
```
Binance client failed to send ${asset?.symbol ?? AssetBNB.symbol} transaction using Ledger
```

File: src/main/api/ledger/binance/transaction.ts [hunk 4]
```diff
@@ -1,53 +1,89 @@
...
+
+    const { result } = await bncClient.transfer(
+      sender,
+      recipient,
+      baseToAsset(amount).amount().toString(),
+      asset ? asset.symbol : AssetBNB.symbol,
+      memo
+    )
+
+    if (result.length === 0 || result[0].hash === undefined) {
```
Comment (veado) [inline src/main/api/ledger/binance/transaction.ts]:
```
result.length === 0 || !result[0].hash
```

File: src/main/api/ledger/binance/transaction.ts [hunk 5]
```diff
@@ -1,53 +1,89 @@
...
+      () => {},
+      () => {},
+      () => {},
+      derivePath
+    )
+
+    if (!sender) {
+      return E.left({
+        errorId: LedgerErrorId.GET_ADDRESS_FAILED,
+        msg: `Getting sender address failed`
```
Comment (veado) [inline src/main/api/ledger/binance/transaction.ts]: `Getting sender address using Ledger failed`

File: src/main/api/ledger/binance/transaction.ts [hunk 6]
```diff
@@ -1,53 +1,89 @@
...
+      })
+    }
+
+    const { result } = await bncClient.transfer(
+      sender,
+      recipient,
+      baseToAsset(amount).amount().toString(),
+      asset ? asset.symbol : AssetBNB.symbol,
```
Comment (veado) [inline src/main/api/ledger/binance/transaction.ts]:
```ts
asset?.symbol ?? AssetBNB.symbol
```

File: src/renderer/services/binance/transaction.ts [hunk 7]
```diff
@@ -1,4 +1,66 @@
...
+    return FP.pipe(
+      Rx.from(window.apiHDWallet.sendLedgerTx(encoded)),
+      RxOp.switchMap(
+        FP.flow(
+          E.fold<LedgerError, TxHash, TxHashLD>(
+            ({ msg }) =>
+              Rx.of(
+                RD.failure({
+                  errorId: ErrorId.SEND_LEDGER_TX,
+                  msg: `Sending Ledger BNB tx failed. (${msg})`
```
Comment (veado) [inline src/renderer/services/binance/transaction.ts]:
```ts
Sending Ledger ${asset?.symbol ?? AssetBNB.symbol} tx failed. (${msg})
```

File: src/renderer/services/thorchain/transaction.ts [hunk 8]
```diff
@@ -102,6 +102,7 @@ export const createTransactionService = (client$: Client$, network$: Network$):
       network,
       asset: params.asset,
       amount: params.amount,
+      sender: params.sender ? params.sender : '',
```
Comment (veado) [inline src/renderer/services/thorchain/transaction.ts]:
```ts
sender: params.sender
```

**Review by veado — CHANGES_REQUESTED** [review]
Nice! Just few (small) things ...

**Review by veado — DISMISSED** [review]
:+1:

File: src/main/api/ledger/binance/transaction.ts [hunk 9]
```diff
@@ -1,53 +1,89 @@
...
+}): Promise<E.Either<LedgerError, TxHash>> => {
+  try {
+    const clientNetwork = toClientNetwork(network)
+    const prefix = getPrefix(clientNetwork)
+    const derivePath = getDerivePath(0)
+
+    const app = new LedgerApp(transport)
+    const client = new Client({ network: clientNetwork })
+
+    await app.showAddress(prefix, derivePath)
```
Comment (veado) [inline src/main/api/ledger/binance/transaction.ts:40]: Do we really need this call of `showAddress`^? By calling `useLedgerSigningDelegate` Binance app on Ledger is showing address + txs address as well

Comment (gromxyz) [reply src/main/api/ledger/binance/transaction.ts:40]: Good catch, removing it 👍

File: src/main/api/ledger/binance/transaction.ts [hunk 10]
```diff
@@ -1,53 +1,87 @@
...
+import { LedgerError, LedgerErrorId, Network } from '../../../../shared/api/types'
+import { toClientNetwork } from '../../../../shared/utils/client'
+import { isError } from '../../../../shared/utils/guard'
 
-export const send = () =>
-  Promise.reject(
-    E.left({
+/**
+ * Sends `MsgSend` message using Ledger
+ */
```
Comment (veado) [inline src/main/api/ledger/binance/transaction.ts]: typo `Sends Binance tx using Ledger`

**Review by veado — APPROVED** [review]

**Pull request merged**